### DISCUSSION TITLE
classes/class_editorscript

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#269] [User-contributed notes]
# classes/class_editorscript

Inherits: RefCounted< Object Base script that can be used to add extension functions to the editor. Description: Scripts extending this class and implementing its_run method can be executed from th...

https://docs.godotengine.org/en/stable/classes/class_editorscript.html

**Comment by misha-cilantro** [comment]
If you're using c# and an external editor, when you try to open the script in the Script window it will open the other editor instead, so you won't be able to run your script. You could change your editor settings, but that's annoying.

A silly but simple fix is to create a sibling GDScript file that loads, instantiates, and runs your C# EditorScript's _Run method. So if I have VerbGenerator.cs and verb_generator.gd in the same folder, this is the code for verb_generator.gd:

```gdscript
@tool
extends EditorScript

func _run():
	var script_node = load("res://tool_scripts/VerbGenerator.cs").new()
	script_node._Run()
```

Now this script you can run from the Script menu, and all it does is load and run your actual C# script.

**Reply by inputrandomness** [reply]
This is a nice workaround for C# but I'm actually editing GDScript code in another editor. I kind of wish there was a simple toggle on the script menu for switching using the built-in editor (or a similar functionality like "Open in Godot Editor" in the file menu).

**Reply by misha-cilantro** [reply]
A "run script" right click option on the file would be cool, too, and could solve it for both languages.

**Reply by Calinou** [reply]
@misha-cilantro See https://github.com/godotengine/godot/pull/93274.

Note that 4.4 has https://github.com/godotengine/godot/pull/96290, which may be preferable to EditorScript for some use cases.

**Reply by StephanGrzelkowski** [reply]
In 4.5 you can add '[GlobalClass]' to your C# script and then you can run your script by script name through the command palette ('ctrlc+shift+p').

**Comment by StephanGrzelkowski** [comment]
I also ran into an issue when I was trying to load a custom resource type with my EditorScript with `GD.Load<MyResourceType>()` which was resulting in a `InvalidCastException`:

`System.InvalidCastException: Unable to cast object of type 'Godot.Resource' to type 'MyResourceType'.`

I solved this by adding `[Tool]` to `MyResourceType`.

**Comment by RobertMic** [comment]
I ran into this while creating an `EditorScript`, if you try to do a `node is MyScript` it will return `false` unless `MyScript` also has a `[Tool]` tag

**Comment by gsisko** [comment]
Anyone know of a good way to debug EditorScripts?
